libbeat/outputs/shipper/shipper_test.go

```diff
@@ -353,48 +353,84 @@ func TestPublish(t *testing.T) {
 
-		batch := outest.NewBatch(events...)
+		client := group.Clients[0].(outputs.NetworkClient)
 
-		// try to publish without the server running
-		err = group.Clients[0].Publish(ctx, batch)
+		err = client.Connect()
 		require.NoError(t, err)
 
+		// Should successfully publish with the server running
+		batch := outest.NewBatch(events...)
+		err = client.Publish(ctx, batch)
+		require.NoError(t, err)
 		expSignals := []outest.BatchSignal{
+			{
+				Tag: outest.BatchACK,
+			},
+		}
+		require.Equal(t, expSignals, batch.Signals)
+
+		stop() // now stop the server and try sending again
+
+		batch = outest.NewBatch(events...) // resetting the batch signals
+		err = client.Publish(ctx, batch)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "failed to publish the batch to the shipper, none of the 2 events were accepted")
+		expSignals = []outest.BatchSignal{
 			{
 				Tag: outest.BatchCancelled, // "cancelled" means there will be a retry without decreasing the TTL
 			},
 		}
 		require.Equal(t, expSignals, batch.Signals)
+		client.Close()
 
-		// Start the server
-		go func() {
-			_ = grpcServer.Serve(listener)
-		}()
+		// Start the server again
+		_, stop = runServer(t, 5, nil, addr)
+		defer stop()
 
-		var actSignals []outest.BatchSignal
+		batch = outest.NewBatch(events...) // resetting the signals
 		expSignals = []outest.BatchSignal{
 			{
 				Tag: outest.BatchACK,
 			},
 		}
 
-		// Poll for the batch to be acknowledged. The gRPC server takes a variable amount
-		// of time to start, so some retries are necessary.
-		require.Eventually(t, func() bool {
-			batch = outest.NewBatch(events...)
-			err = group.Clients[0].Publish(ctx, batch)
-			require.NoError(t, err)
-
-			actSignals = batch.Signals
-			return reflect.DeepEqual(expSignals, batch.Signals)
-		}, 5*time.Second, 500*time.Millisecond)
+		// The backoff wrapper should take care of the errors and
+		// retries while the server is still starting
+		err = client.Connect()
+		require.NoError(t, err)
 
-		// Use require.Equal() on the final signal set. If the Eventually() loop above
-		// failed this will print the difference between the signal sets.
-		require.Equal(t, expSignals, actSignals)
+		err = client.Publish(ctx, batch)
+		require.NoError(t, err)
+		require.Equal(t, expSignals, batch.Signals)
 	})
 }
 
+// runServer mocks the shipper mock server for testing
+// `qSize` is a sice of the event buffer in the mock
+// `err` is a preset error that the server will serve to the client
+// `listenAddr` is the address for the server to listen
+// returns `actualAddr` where the listener actuall is and the `stop` function to stop the server
+func runServer(t *testing.T, qSize int, err error, listenAddr string) (actualAddr string, stop func()) {
+	producer := sc.NewProducerMock(qSize)
+	producer.Error = err
+	grpcServer := grpc.NewServer()
+	sc.RegisterProducerServer(grpcServer, producer)
+
+	listener, err := net.Listen("tcp", listenAddr)
```

Comment:
There is a way to avoid using ports at all, gRPC has an in-memory buffer implementation for `net.Conn`: https://pkg.go.dev/google.golang.org/grpc/test/bufconn

For an example: https://stackoverflow.com/a/52080545

Reply:
@cmacknz this would require to change the client implementation. The `Connect()` function would look like this:

```golang
conn, err := grpc.DialContext(ctx, "bufnet", grpc.WithContextDialer(bufDialer), grpc.WithInsecure())
if err != nil {
    t.Fatalf("Failed to dial bufnet: %v", err)
}
defer conn.Close()
client := pb.NewGreeterClient(conn)
```

instead of https://github.com/elastic/beats/pull/32013/files#diff-72f0b30fee9226d221a79a223ccddc9ba33df51ab286aa8223a89943d8a22247R107-R114